vulkan/vulkan_swapchain.cc

```diff
@@ -0,0 +1,511 @@
+
+    if (!vulkan_image->IsValid()) {
+      return false;
+    }
+
+    images_.emplace_back(std::move(vulkan_image));
+
+    // Populate the surface.
+    auto surface = CreateSkiaSurface(skia_context, image, surface_size);
+
+    if (surface == nullptr) {
+      return false;
+    }
+
+    surfaces_.emplace_back(std::move(surface));
+  }
+
+  FTL_DCHECK(backbuffers_.size() == images_.size());
+  FTL_DCHECK(images_.size() == surfaces_.size());
+
+  return true;
+}
+
+VulkanBackbuffer* VulkanSwapchain::NextBackbuffer() {
+  auto available_backbuffers = backbuffers_.size();
+
+  if (available_backbuffers == 0) {
+    return nullptr;
+  }
+
+  auto next_backbuffer_index =
+      (current_backbuffer_index_ + 1) % backbuffers_.size();
+
+  auto& backbuffer = backbuffers_[next_backbuffer_index];
+
+  if (!backbuffer->IsValid()) {
+    return nullptr;
+  }
+
+  current_backbuffer_index_ = next_backbuffer_index;
+  return backbuffer.get();
+}
+
+VulkanSwapchain::AcquireResult VulkanSwapchain::AcquireSurface() {
+  AcquireResult error = {AcquireType::ErrorSurfaceLost, nullptr};
+
+  if (!IsValid()) {
+    return error;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Step 0:
+  // Acquire the next available backbuffer.
+  // ---------------------------------------------------------------------------
+  auto backbuffer = NextBackbuffer();
+
+  if (backbuffer == nullptr) {
+    return error;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Step 1:
+  // Wait for use readiness.
+  // ---------------------------------------------------------------------------
+  if (!backbuffer->WaitFences()) {
+    return error;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Step 2:
+  // Put semaphores in unsignled state.
+  // ---------------------------------------------------------------------------
+  if (!backbuffer->ResetFences()) {
+    return error;
+  }
+
+  // ---------------------------------------------------------------------------
+  // Step 3:
+  // Acquire the next image index.
+  // ---------------------------------------------------------------------------
+  uint32_t next_image_index = 0;
+
+  switch (VK_CALL_LOG_ERROR(
+      vk.AcquireNextImageKHR(device_.Handle(),                      //
+                             swapchain_,                            //
+                             std::numeric_limits<uint64_t>::max(),  //
+                             backbuffer->UsageSemaphore(),          //
+                             VK_NULL_HANDLE,                        //
+                             &next_image_index))                    //
+          ) {
+    case VK_SUCCESS:
+      break;
+    case VK_ERROR_OUT_OF_DATE_KHR:
+      return {AcquireType::ErrorSurfaceOutOfDate, nullptr};
+    default:
+      // Includes VK_ERROR_SURFACE_LOST_KHR.
+      return {AcquireType::ErrorSurfaceLost, nullptr};
```

Comment:
Maybe have an explicit case for VK_ERROR_SURFACE_LOST_KHR.  The default case could still return AcquireType::ErrorSurfaceLost (and maybe also log that it encountered an unexpected case).